addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CommandInjectionScanRule.java

```diff
@@ -592,80 +595,83 @@ private boolean testCommandInjection(
-                            .setOtherInfo(otherInfo)
-                            .raise();
-
-                    // All done. No need to look for vulnerabilities on subsequent
-                    // payloads on the same request (to reduce performance impact)
-                    return true;
-                }
-
-            } catch (IOException ex) {
-                // Do not try to internationalise this.. we need an error message in any event..
+                // use TimingUtils to detect a response to sleep payloads
+                isInjectable =
+                        TimingUtils.checkTimingDependence(
+                                BLIND_REQUEST_LIMIT, BLIND_SECONDS_LIMIT,
+                                requestSender,
+                                TIME_CORRELATION_ERROR_RANGE,
+                                TIME_SLOPE_ERROR_RANGE);
+            } catch (SocketException e) {
+                log.debug(
+                        "Caught {} {} when accessing: {}.\n The target may have replied with a poorly formed redirect due to our input.",
+                        e.getClass().getName(),
+                        e.getMessage(),
+                        getBaseMsg().getRequestHeader().getURI());
+                // Something went wrong, move to next blind iteration
+                continue;
+            } catch (IOException e) {
+                // Do not try to internationalise this... we need an error message in any
+                // event...
                 // if it's in English, it's still better than not having it at all.
                 log.warn(
                         "Blind Command Injection vulnerability check failed for parameter [{}] and payload [{}] due to an I/O error",
                         paramName,
-                        payload,
-                        ex);
+                        paramValue,
+                        e);
+            }
+
+            if (isInjectable) {
+                // We Found IT!
+                // First do logging
+                log.debug(
+                        "[Blind OS Command Injection Found] on parameter [{}] with value [{}]",
+                        paramName,
+                        paramValue);
+                String otherInfo = getOtherInfo("time-based", paramValue);
+
+                // raise the alert directly with MEDIUM Confidence...
+                newAlert()
+                        .setConfidence(Alert.CONFIDENCE_MEDIUM)
+                        .setParam(paramName)
+                        .setAttack(paramValue)
+                        // just attach this alert to the last sent message
+                        .setMessage(message.get())
+                        .setOtherInfo(otherInfo)
+                        .raise();
+
+                // All done. No need to look for vulnerabilities on subsequent
+                // payloads on the same request (to reduce performance impact)
+                return true;
             }
```

Comment:
This can be moved back to its place, and some of the changes done in the catch block reverted.

Reply:
I think I addressed this properly by reintroducing the double try-catch block and putting the success case nested inside of the outer try-catch, but I'm a bit uncertain, so please take a look and let me know if you notice any problems.

Reply:
There are some more reverts that could be done, added as new comments.